source/custom/nidaqmx_service.custom.cpp

```diff
@@ -1,48 +1,97 @@
+  NiDAQmxService* service_;
+};
+
+void NiDAQmxService::register_async_functions(::grpc::ServerCompletionQueue* completion_queue)
+{
+  DoneEventCallbackOperation::register_completion_queue_element(this, completion_queue);
+}
+
 //---------------------------------------------------------------------
 //---------------------------------------------------------------------
 ::grpc::Status
-NiDAQmxService::RegisterDoneEvent(::grpc::ServerContext* context, const RegisterDoneEventRequest* request, ::grpc::ServerWriter<RegisterDoneEventResponse>* writer)
+NiDAQmxService::process_RegisterDoneEvent(const RegisterDoneEventMethodContextPtr& async_method_context)
 {
-  if (context->IsCancelled()) {
-    return ::grpc::Status::CANCELLED;
-  }
+  //if (context->IsCancelled()) {
+  //  return ::grpc::Status::CANCELLED;
+  //}
+  auto& request = async_method_context->request;
+  auto& writer = async_method_context->writer;
   try {
-    auto task_grpc_session = request->task();
+    auto task_grpc_session = request.task();
     TaskHandle task = session_repository_->access_session(task_grpc_session.id(), task_grpc_session.name());
-    uInt32 options = request->options();
+    uInt32 options = request.options();
 
-    auto registration = DoneEventCallbackRouter::register_handler(
-        [writer](TaskHandle task, int32 callback_status) {
+    writer.SendInitialMetadata(nidevice_grpc::NO_TAG);
+
+    async_method_context->registration = DoneEventCallbackRouter::register_handler(
+        [async_method_context]  // Copy the shared_ptr to ensure the callback has access to the writer.
+        (TaskHandle task, int32 callback_status) {
           RegisterDoneEventResponse callback_response;
           callback_response.set_status(callback_status);
-          writer->Write(callback_response);
+          async_method_context->writer.Write(callback_response, nidevice_grpc::NO_TAG);
           return DAQmxSuccess;
         });
 
-    auto status = library_->RegisterDoneEvent(task, options, DoneEventCallbackRouter::handle_callback, registration.token());
+    auto status = library_->RegisterDoneEvent(task, options, DoneEventCallbackRouter::handle_callback, async_method_context->registration->token());
 
     if (status) {
       RegisterDoneEventResponse failed_to_register_response;
       failed_to_register_response.set_status(status);
-      writer->Write(failed_to_register_response);
+      writer.Write(
+          failed_to_register_response,
+          nidevice_grpc::NO_TAG);
     }
 
-    while (!context->IsCancelled()) {
-      std::this_thread::sleep_for(std::chrono::milliseconds(10));
-    }
+    //while (!context->IsCancelled()) {
+    //  std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    //}
     return ::grpc::Status::OK;
   }
   catch (nidevice_grpc::LibraryLoadException& ex) {
-    return ::grpc::Status(::grpc::NOT_FOUND, ex.what());
+    writer.Finish(::grpc::Status(::grpc::NOT_FOUND, ex.what()), nidevice_grpc::NO_TAG);
```

Comment:
TODO: I think this is the right idea, but I'm not sure it works with the rest of my code.